### DISCUSSION TITLE
Trouble installing Pi-hole on UDM Special Edition, please help

### DISCUSSION THREAD

**Discussion by GigaBypass9876** [unifi-utilities/unifi-common#376] [Q&A]
Thanks to all who have contributed to this excellent set of tools. I have run into a few problems with Pi-hole instructions and had a few questions on getting the dns programs running. 
I am installing on UDM SE with up to date f/w.
1. When I executed 
/mnt/data/on_boot.d/05-install-cni-plugins.sh
I got the following error:
/mnt/data/on_boot.d/05-install-cni-plugins.sh: line 8: syntax error near unexpected token `newline' /mnt/data/on_boot.d/05-install-cni-plugins.sh: line 8: `'

Q: Does this mean it did not run successfully?

2. I executed this to copy the 20-dns file:
curl -L https://github.com/boostchicken-dev/udm-utilities/raw/master/cni-plugins/20-dns.conflist -o /mnt/data/podman/cni/dns.conflist

I edited this for my specific IP and vlan, but was not sure if the file was supposed to named dns-conflist (as in the instructions) or 20-dns.conflist. I copied the file and have both dns.conflist and 20-dns.conflist in the directory.

3. for this command:
curl -L https://github.com/boostchicken-dev/udm-utilities/raw/master/cni-plugins/20-dns.conflist -o /etc/cni/net.d/dns.conflist

I get an error that the directory does not exist. Indeed it does not. There are many directories within /etc/ but cni is not among them. I even searched for net.d to see if there was a typo, and nothing.

4. Finally, installing podman for me involved TWO unzip commands because the first (unzip udmse-podman-install.zip) unzipped a file which was a zip file (podman-install.zip), which then needed to be unzipped. I discovered this because entering 
podman network create dns
resulted in an error, “podman: command not found”
after investigating I saw the 2nd zip file and unzipped it, but I still get this error even after the second unzip created and inflated the various files it was supposed to.
Q: How do I get podman to install correctly?

Thanks in advance!

**Comment by gatesry** [comment]
Hello, see my answers below. 

1. Make sure your end of line sequencing is LF and CRLF. This will prevent the script from executing.
2. Naming does not matter, as long as it ends in `.conflist` 
3. If the `05-install-cni-plugins.sh` script does not run, it will not generate the correct directories. 
4. I don't have an SE variant, but I would check the Issues and previous Discussions to see if a solution is found.

**Reply by GigaBypass9876** [reply]
Also, the reason this command cp /mnt/data/podman/cni/20-dns.conflist /etc/cni/net.d/dns.conflist produced the error: no file or directory is because the unzipped podman files are located at /root/etc/cni/, root/etc/containers and root/etc/usr. In other words, the files are in a directory called root in the root directory. is that how it is supposed to be? wouldn't the line above need to be 
cp /mnt/data/podman/cni/20-dns.conflist /root/etc/cni/net.d/dns.conflist 
The error is why I physically created /etc/cni/net.d not realizing this other root directory.  
I suspect something glitched when I was installing these. If someone could please weigh in, I would appreciate it.

**Reply by gatesry** [reply]
Have you taken at look at #321, specifically this [comment](https://github.com/boostchicken-dev/udm-utilities/issues/321#issuecomment-1098460030)?

**Reply by GigaBypass9876** [reply]
The issue in #321 seems to be 10-dns running before containers are ready. My problem is 10-dns cannot run whatsoever because the podman command is not found (because something is amiss with the install on a UDM SE). For those that have podman running on a UDM or ideally UDM SE, where exactly are the usr and cni directories? somehow mine were created in a directory named /root/, not The root directory (i.e., /root/etc/cni/net.d and not etc/cni/net.d). so when I try to run 10-dns, it returned an error 'directory not found'. So, I created them and copied dns.conflist to them. it still fails with file not found. any insights on this and possibly a way to fix?

**Reply by gatesry** [reply]
Strange! I'm pinging @mazzy89, who also has a UDM-SE. He might be able to add some color.

**Reply by mazzy89** [reply]
> where exactly are the usr and cni directories

@GigaBypass9876 I would recommend you to take a look at this article https://www.howtogeek.com/117435/htg-explains-the-linux-directory-structure-explained/

UDM SE runs Debian so in order to make this works you need to be familiar with Linux directories structure.

That's said Podman artifact produced by the built once decompressed contains files that are located in the relative directories of how files in the UDM-SE filesystem should be placed starting from the root which in Linux is not `/root` but `/`. 

In a common Linux system `/root` is the home dir of the root user and not the root of the fs.

**Reply by GigaBypass9876** [reply]
I must have inadvertently copied the podman zipfile to the /root user directory and unzipped it there, creating /root/etc and /root/cni directories in the process. When I discovered this, I copied the zipfile to the / <root> system directory and unzipped it there. Is there any reason why this would not create the directories /etc/cni/net.d? It did not, so I created them myself to eliminate the "directory not found" error. I guess the podman unzip needs to be completed before running 05-install-cni-plugins.sh which uses those /etc directories podman creates. Is that right? How do I undo what I have done because as I mentioned, unzipping podman in the root system directory did not create those /etc/cni and /etc/usr directories. Should I delete /root/etc/cni and /root/etc/usr directories and unzip podman once more from <root> system directory? I appreciate your responses in this as I am less familiar with Linux.

**Reply by mazzy89** [reply]
>  I guess the podman unzip needs to be completed before

Podman is an **hard** requirement in order to use CNI. CNI stands for Container Network Interface. Podman manages containers. Without Podman, no CNI would ever work.


> Should I delete /root/etc/cni and /root/etc/usr directories and unzip podman once more from system directory?

yes.

> I am less familiar with Linux.

You are trying to manage a router that runs Debian. I really recommend you get familiar with Linux. The number of things that will go wrong, especially during upgrade/reboot etc... if not managed properly are limitless.

**Reply by GigaBypass9876** [reply]
Thank you all for the help on this.
I am almost there and have the pihole_dote.sh file customized and ready to go but encountered these
two errors when creating the dns:

root@UDM-SE:~# podman network create dns
ERRO[0000] Failed to built-in GetDriver graph btrfs /var/lib/containers/storage 
Error: the network name dns is already used

I did come across issue #272 where Mazzy89 encountered this error and solved by creating a /etc/containers/storage.conf file. Could you share some of the lines from that file?  Thanks

**Reply by gatesry** [reply]
@GigaBypass9876 The config files are referenced in #321

**Reply by GigaBypass9876** [reply]
Many thanks! I have incorporated the storage.conf file and now I get one error:
root@UDM-SE:~# podman network create dns
Error: the network name dns is already used

Any thoughts on what could be causing this?

**Reply by mazzy89** [reply]
Isn't it very clear the message? 🤷‍♂️ You are trying to create a network that already exist. Delete it first and create it again. Another recommendation. Get familiar with the podman CLI and you will realize that there is a way to list network. Then you will understand that there is a network with dns name already there and after that finally you will understand that is time to delete that network before create a new one with the same name

**Reply by GigaBypass9876** [reply]
I really appreciate your patience with this. I am just attempting to get the pihole adblocker running on a new (and totally unfamiliar) UDM SE after replacing my USG running dnsmasq via json file for several years. The ads are brutal and there is a lot of learning curve, so I appreciate any wisdom this forum can provide.  

I encountered an error creating the pihole container and was reading up on possible causes. In the process I ran ‘systemctl start udm-boot’ and realized it was not running. That install went without error, so this was surprising. The output is below. So before going further, I need to tackle this. I have tried to install with both 
'curl -fsL "https://raw.githubusercontent.com/boostchicken-dev/udm-utilities/HEAD/on-boot-script/remote_install.sh" | /bin/shI'
and with 
‘curl -L https://udm-boot.boostchicken.dev -o udm-boot_1.0.5_all.deb’ 
When I check status to see if it is running, I encounter these errors.
Thanks in advance for any insights.

root@UDM-SE:~# systemctl status udm-boot
 udm-boot.service - Run On Startup UDM
   Loaded: loaded (/etc/systemd/system/udm-boot.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Fri 2022-06-17 08:11:44 EDT; 1min 3s ago
  Process: 20428 ExecStart=/bin/bash -c mkdir -p /mnt/data/on_boot.d && find -L /mnt/data/on_
      CPU: 1.934s

Jun 17 08:11:44 UDM-SE bash[20428]: /mnt/data/on_boot.d/pi_hole_update.sh: line 29: --hostnam
Jun 17 08:11:44 UDM-SE bash[20428]: /mnt/data/on_boot.d/pi_hole_update.sh: line 30: -e: comma
Jun 17 08:11:44 UDM-SE bash[20428]: /mnt/data/on_boot.d/pi_hole_update.sh: line 31: -e: comma
Jun 17 08:11:44 UDM-SE bash[20428]: /mnt/data/on_boot.d/pi_hole_update.sh: line 32: -e: comma
Jun 17 08:11:44 UDM-SE bash[20428]: /mnt/data/on_boot.d/pi_hole_update.sh: line 33: -e: comma
Jun 17 08:11:44 UDM-SE bash[20428]: [96B blob data]
Jun 17 08:11:44 UDM-SE systemd[1]: udm-boot.service: Control process exited, code=exited, sta
Jun 17 08:11:44 UDM-SE systemd[1]: udm-boot.service: Failed with result 'exit-code'.
Jun 17 08:11:44 UDM-SE systemd[1]: Failed to start Run On Startup UDM.
Jun 17 08:11:44 UDM-SE systemd[1]: udm-boot.service: Consumed 1.934s CPU time.
lines 1-16/16 (END)
root@UDM-SE:~# journalctl -xe
Jun 17 08:18:48 UDM-SE systemd[1]: udm-boot.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- The unit udm-boot.service has entered the 'failed' state with result 'exit-code'.
Jun 17 08:18:48 UDM-SE systemd[1]: Failed to start Run On Startup UDM.
-- Subject: A start job for unit udm-boot.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A start job for unit udm-boot.service has finished with a failure.
-- 
-- The job identifier is 919 and the job result is failed.
Jun 17 08:18:48 UDM-SE systemd[1]: udm-boot.service: Consumed 1.845s CPU time.
-- Subject: Resources consumed by unit runtime
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- The unit udm-boot.service completed and consumed the indicated resources.
Jun 17 08:19:19 UDM-SE -bash[19996]: HISTORY: PID=19996 UID=0 systemctl status udm-boot.servi
Jun 17 08:19:31 UDM-SE -bash[19996]: HISTORY: PID=19996 UID=0 "journalctl -xe
Jun 17 08:19:47 UDM-SE -bash[19996]: HISTORY: PID=19996 UID=0 journalctl -xe
lines 1137-1159/1159 (END) 
root@UDM-SE:~# curl -L https://udm-boot.boostchicken.dev -o udm-boot_1.0.5_all.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  3364  100  3364    0     0   9805      0 --:--:-- --:--:-- --:--:-- 3285k
root@UDM-SE:~# dpkg -i udm-boot_1.0.5_all.deb
Selecting previously unselected package udm-boot.
(Reading database ...  (Reading database ... 5% (Reading database ... 10% (Reading database ... 15% (Reading database ... 20% (Reading database ... 25% (Reading database ... 30% (Reading database ... 35% (Reading database ... 40% (Reading database ... 45% (Reading database ... 50% (Reading database ... 55% (Reading database ... 60% (Reading database ... 65% (Reading database ... 70% (Reading database ... 75% (Reading database ... 80% (Reading database ... 85% (Reading database ... 90% (Reading database ... 95%
(Reading database ... 100%
(Reading database ... 73417 files and directories currently installed.)
Preparing to unpack udm-boot_1.0.5_all.deb ...
/var/lib/dpkg/tmp.ci/preinst: 19: /var/lib/dpkg/tmp.ci/preinst: /sbin/ssh-proxy: not found
dpkg: error processing archive udm-boot_1.0.5_all.deb (--install):
 subprocess new pre-installation script returned error exit status 127
/var/lib/dpkg/tmp.ci/postrm: 24: /var/lib/dpkg/tmp.ci/postrm: /sbin/ssh-proxy: not found
dpkg: error while cleaning up:
 subprocess new post-removal script returned error exit status 127
Errors were encountered while processing:
 udm-boot_1.0.5_all.deb
